File: frontend/src/app/types/accounts.ts
import type { ArchiveItem } from './archives';
import type { Contact, Group, MessageTask } from './messaging';
import type { OperationTarget, OperationTaskAttempt } from './operations';

export type LoginFlow = {
  id: number;
  account_id: number;
  method: string;
  status: string;
  code_preview: string | null;
  code_expires_at: string | null;
  qr_payload: string | null;
  created_at: string;
};

export type AccountLoginForm = {
  account: Account | null;
  step: 'method' | 'code' | 'qr' | 'password';
  method: 'code' | 'qr';
  code: string;
  password_2fa: string;
  flow: LoginFlow | null;
  error: string;
};

export type Account = {
  id: number;
  pool_id: number | null;
  pool_name: string;
  display_name: string;
  username: string | null;
  tg_first_name: string;
  tg_last_name: string;
  tg_bio: string;
  avatar_object_key: string;
  avatar_preview_url: string;
  profile_sync_status: string;
  profile_sync_error: string;
  profile_synced_at: string | null;
  phone_masked: string;
  phone_number: string | null;
  status: string;
  health_score: number;
  last_active_at: string | null;
  deleted_at?: string | null;
  deleted_by?: string;
  delete_reason?: string;
  developer_app_id: number | null;
  developer_app_name: string | null;
  developer_api_id: number | null;
  developer_app_health_status: string | null;
  developer_app_version: number;
  proxy_id: number | null;
  proxy_name: string | null;
  proxy_local_address: string | null;
  proxy_status: string | null;
  proxy_alert_status: string | null;
};

export type AccountAvailabilitySummary = {
  id: string;
  tenant_id: number;
  account_id: number;
  send_available: boolean;
  listen_available: boolean;
  join_available: boolean;
  comment_available: boolean;
  profile_available: boolean;
  code_read_available: boolean;
  remaining_capacity: number;
  unavailable_reason: string;
  next_retry_at: string | null;
  failure_trend: Record<string, number>;
  health_score: number;
  risk_level: string;
  score_reasons: string[];
  non_score_reasons: string[];
  updated_at: string;
};

export type AccountPool = {
  id: number;
  tenant_id: number;
  name: string;
  description: string;
  is_default: boolean;
  account_count: number;
};

export type DeveloperApp = {
  id: number;
  app_name: string;
  api_id: number;
  is_active: boolean;
  health_status: string;
  max_accounts: number;
  assigned_accounts: number;
  credentials_version: number;
  last_assigned_at: string | null;
  last_check_at: string | null;
  last_error: string;
  notes: string;
  created_at: string;
  updated_at: string;
};

export type VerificationCode = {
  id: number;
  account_id: number;
  source: string;
  code_preview: string | null;
  expires_at: string | null;
  viewed_by: string;
  viewed_at: string | null;
  status: string;
  raw_hint: string;
  created_at: string;
};

export type AccountSyncRecord = {
  id: number;
  account_id: number;
  sync_type: string;
  trigger_source: string;
  status: string;
  result_count: number;
  failure_type: string;
  failure_detail: string;
  scheduled_at: string;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
};

export type AccountSecurityPreviewItem = {
  account_id: number;
  account_name: string;
  phone_masked: string;
  phone_number?: string | null;
  session_status: string;
  trusted_session_status: string;
  external_authorization_count: number;
  two_fa_status: string;
  profile_status: string;
  generated_display_name: string;
  generated_first_name: string;
  generated_last_name: string;
  generated_bio: string;
  username_candidates: string[];
  avatar_source: string;
  precheck_status: string;
  blockers: string[];
  warnings: string[];
  suggested_actions: string[];
};

export type AccountSecurityPrecheck = {
  batch_preview_id: string;
  summary: Record<string, number>;
  items: AccountSecurityPreviewItem[];
  action_types: string[];
  trace_id: string;
};

export type AccountSecurityBatchItem = {
  id: number;
  batch_id: number;
  tenant_id: number;
  account_id: number;
  status: string;
  precheck_status: string;
  cleanup_status: string;
  two_fa_status: string;
  profile_status: string;
  username_status: string;
  avatar_status: string;
  external_devices_before: number;
  external_devices_after: number;
  generated_display_name: string;
  generated_first_name: string;
  generated_last_name: string;
  generated_bio: string;
  generated_username: string;
  username_candidates: string[];
  avatar_source: string;
  skipped_reason: string;
  failure_type: string;
  failure_detail: string;
  next_retry_at: string | null;
  trace_id: string;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
};

export type AccountSecurityBatch = {
  id: number;
  tenant_id: number;
  action_types: string[];
  status: string;
  total_count: number;
  success_count: number;
  skipped_count: number;
  failed_count: number;
  created_by: string;
  confirmed_by: string;
  confirm_text: string;
  password_strategy: string;
  profile_strategy: Record<string, unknown>;
  username_strategy: Record<string, unknown>;
  avatar_strategy: Record<string, unknown>;
  overwrite_existing_profile: boolean;
  reason: string;
  trace_id: string;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
  items: AccountSecurityBatchItem[];
};

export type AccountAuthorizationSnapshot = {
  id: number;
  account_id: number;
  batch_id: number | null;
  authorization_hash_ciphertext: string;
  is_platform_trusted: boolean;
  is_current_session: boolean;
  device_model: string;
  platform: string;
  system_version: string;
  api_id: number;
  app_name: string;
  app_version: string;
  ip_masked: string;
  country: string;
  region: string;
  date_created: string | null;
  date_active: string | null;
  status: string;
  scanned_at: string;
};

export type AccountSecuritySnapshot = {
  id: number;
  account_id: number;
  trusted_session_status: string;
  two_fa_status: string;
  external_authorization_count: number;
  last_device_scan_at: string | null;
  last_2fa_check_at: string | null;
  profile_status: string;
  profile_last_updated_at: string | null;
  trusted_device_label: string;
  last_hardened_at: string | null;
  last_error: string;
  trace_id: string;
  created_at: string;
  updated_at: string;
};

export type AccountSecurityDetail = {
  account_id: number;
  snapshot: AccountSecuritySnapshot;
  authorizations: AccountAuthorizationSnapshot[];
  recent_batches: AccountSecurityBatch[];
};

export type VerificationTask = {
  id: number;
  tenant_id: number;
  account_id: number | null;
  group_id: number | null;
  message_task_id: number | null;
  verification_type: string;
  detected_reason: string;
  suggested_action: string;
  target_peer_id: string;
  target_display: string;
  requires_user_confirm: boolean;
  status: string;
  failure_detail: string;
  created_at: string;
  handled_at: string | null;
  issue_scope: 'account' | 'target' | string;
  issue_category: 'account_restricted' | 'group_restriction' | 'verification' | string;
  can_auto_resolve: boolean;
  requires_target_recheck: boolean;
  resolution_entry_label: string;
};

export type AccountCloneItem = {
  id: number;
  plan_id: number;
  source_account_id: number;
  target_account_id: number;
  target_type: string;
  target_peer_id: string;
  target_display: string;
  status: string;
  failure_type: string;
  failure_detail: string;
  created_at: string;
  executed_at: string | null;
};

export type AccountClonePlan = {
  id: number;
  source_account_id: number;
  target_account_id: number | null;
  target_account_ids: number[];
  target_accounts_summary: Array<{ id: number; display_name: string; status: string; items_total: number; items_done: number; items_failed: number }>;
  clone_scope: string;
  status: string;
  items_total: number;
  items_done: number;
  items_failed: number;
  failure_detail: string;
  created_by: string;
  created_at: string;
  confirmed_at: string | null;
  items: AccountCloneItem[];
  items_by_target: Record<string, AccountCloneItem[]>;
};

export type AccountGroup = Group & {
  permission_label: string;
  account_can_send: boolean;
  last_sent_at: string | null;
};

export type AccountRiskDiagnostic = {
  level: string;
  code: string;
  title: string;
  detail: string;
  source: string;
  action: string;
  occurred_at: string | null;
};

export type ManualOperationRecord = {
  id: number;
  tenant_id: number;
  account_id: number;
  target_id: number | null;
  operation_type: string;
  content: string;
  status: string;
  failure_type: string;
  failure_detail: string;
  remote_message_id: string;
  actor: string;
  created_at: string;
};

export type ProfileSyncRecord = {
  id: number;
  account_id: number;
  actor: string;
  before_snapshot: string;
  after_snapshot: string;
  avatar_object_key: string;
  status: string;
  failure_type: string;
  failure_detail: string;
  remote_detail: string;
  created_at: string;
  synced_at: string | null;
};

export type AccountDetail = {
  account: Account;
  risk_diagnostics: AccountRiskDiagnostic[];
  login_flows: LoginFlow[];
  verification_codes: VerificationCode[];
  profile_sync_records: ProfileSyncRecord[];
  sync_records: AccountSyncRecord[];
  next_sync_at: string | null;
  sync_due: boolean;
  sync_status_text: string;
  contacts: Contact[];
  groups: AccountGroup[];
  operation_targets: OperationTarget[];
  message_records: MessageTask[];
  manual_operation_records: ManualOperationRecord[];
  operation_task_attempts: OperationTaskAttempt[];
  clone_plans: AccountClonePlan[];
  verification_tasks: VerificationTask[];
  stats: Record<string, number>;
};

export type AccountPoolDetail = {
  pool: AccountPool;
  accounts: Account[];
  contacts: Contact[];
  verification_tasks: VerificationTask[];
  clone_plans: AccountClonePlan[];
  message_records: MessageTask[];
  stats: Record<string, number>;
};

export type GroupDetail = {
  group: Group;
  accounts: Array<{
    id: number;
    display_name: string;
    username: string | null;
    status: string;
    health_score: number;
    permission_label: string;
    can_send: boolean;
    is_listener?: boolean;
    last_sent_at: string | null;
  }>;
  listener_accounts: Array<{ id: number; display_name: string; username: string | null; status: string }>;
  recent_context_messages: Array<{
    id: number;
    listener_account_id: number;
    sender_name: string;
    content: string;
    message_type: string;
    sent_at: string | null;
    used_for_ai: boolean;
  }>;
  recent_campaigns: Array<Record<string, unknown>>;
  recent_archives: ArchiveItem[];
  verification_tasks: VerificationTask[];
  stats: Record<string, number>;
};

File: backend/app/schemas/accounts.py
from __future__ import annotations
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

from .api import ApiModel
from .campaigns import MessageTaskOut
from .groups import VerificationTaskOut
from .operations import ManualOperationRecordOut, OperationTargetOut, OperationTaskAttemptOut


# ── Request schemas ──

class TgAccountCreate(BaseModel):
    tenant_id: int = 1
    pool_id: int | None = None
    display_name: str
    username: str | None = None
    phone_masked: str | None = None
    phone_number: str | None = None


class AccountPoolCreate(BaseModel):
    tenant_id: int = 1
    name: str
    description: str = ""
    is_default: bool = False


class AccountPoolUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    is_default: bool | None = None


class MoveAccountPoolRequest(BaseModel):
    pool_id: int


class TgAccountProfileUpdate(BaseModel):
    display_name: str
    tg_first_name: str = ""
    tg_last_name: str = ""
    tg_bio: str = Field(default="", max_length=220)
    avatar_object_key: str = ""


class LoginStartRequest(BaseModel):
    method: str = Field(pattern="^(code|qr)$")
    force: bool = False


class LoginVerifyRequest(BaseModel):
    code: str | None = None
    password_2fa: str | None = None


class SensitiveActionReasonRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reason: str = Field(min_length=1, max_length=255)

    @model_validator(mode="after")
    def normalize_reason(self) -> "SensitiveActionReasonRequest":
        self.reason = self.reason.strip()
        if not self.reason:
            raise ValueError("操作原因不能为空")
        return self


class AccountClonePlanCreate(BaseModel):
    tenant_id: int = 1
    source_account_id: int
    target_account_id: int | None = None
    target_account_ids: list[int] = Field(default_factory=list)
    clone_scope: list[str] = Field(default_factory=lambda: ["contacts", "groups"])


# ── Output schemas ──

class AvatarUploadOut(BaseModel):
    object_key: str
    preview_url: str
    content_type: str
    size: int


class AccountOut(ApiModel):
    id: int
    tenant_id: int
    pool_id: int | None = None
    pool_name: str = "默认账号池"
    display_name: str
    username: str | None
    tg_first_name: str = ""
    tg_last_name: str = ""
    tg_bio: str = ""
    avatar_object_key: str = ""
    avatar_preview_url: str = ""
    profile_sync_status: str = "未同步"
    profile_sync_error: str = ""
    profile_synced_at: datetime | None = None
    phone_masked: str
    phone_number: str | None = None
    status: str
    health_score: float
    last_active_at: datetime | None
    created_at: datetime
    developer_app_id: int | None
    developer_app_name: str | None = None
    developer_api_id: int | None = None
    developer_app_health_status: str | None = None
    developer_app_version: int
    proxy_id: int | None = None
    proxy_name: str | None = None
    proxy_local_address: str | None = None
    proxy_status: str | None = None
    proxy_alert_status: str | None = None
    deleted_at: datetime | None = None
    deleted_by: str = ""
    delete_reason: str = ""


class AccountPoolOut(ApiModel):
    id: int
    tenant_id: int
    name: str
    description: str
    is_default: bool
    account_count: int = 0
    created_at: datetime
    updated_at: datetime


class LoginFlowOut(ApiModel):
    id: int
    account_id: int
    method: str
    status: str
    code_preview: str | None
    code_expires_at: datetime | None
    qr_payload: str | None
    created_at: datetime


class VerificationCodeOut(ApiModel):
    id: int
    account_id: int
    source: str
    code_preview: str | None
    expires_at: datetime | None
    viewed_by: str
    viewed_at: datetime | None
    status: str
    raw_hint: str
    created_at: datetime


class ProfileSyncRecordOut(ApiModel):
    id: int
    tenant_id: int
    account_id: int
    actor: str
    before_snapshot: str
    after_snapshot: str
    avatar_object_key: str
    status: str
    failure_type: str
    failure_detail: str
    remote_detail: str
    created_at: datetime
    synced_at: datetime | None


class AccountSyncRecordOut(ApiModel):
    id: int
    tenant_id: int
    account_id: int
    sync_type: str
    trigger_source: str
    status: str
    result_count: int
    failure_type: str
    failure_detail: str
    scheduled_at: datetime
    started_at: datetime | None
    finished_at: datetime | None
    created_at: datetime


class ContactOut(ApiModel):
    id: int
    tenant_id: int
    account_id: int
    peer_id: str
    display_name: str
    username: str | None
    phone_masked: str
    phone_number: str | None = None
    contact_type: str
    is_mutual: bool
    last_message_at: datetime | None
    last_synced_at: datetime
    created_at: datetime


class AccountGroupOut(ApiModel):
    id: int
    tenant_id: int
    tg_peer_id: str
    title: str
    group_type: str
    member_count: int
    auth_status: str
    can_send: bool
    active_window: str
    daily_limit: int
    account_cooldown_seconds: int
    group_cooldown_seconds: int
    topic_direction: str
    banned_words: str
    link_whitelist: str
    require_review: bool
    permission_label: str = "普通成员"
    account_can_send: bool = True
    last_sent_at: datetime | None = None


class AccountCloneItemOut(ApiModel):
    id: int
    tenant_id: int
    plan_id: int
    source_account_id: int
    target_account_id: int
    target_type: str
    target_peer_id: str
    target_display: str
    status: str
    failure_type: str
    failure_detail: str
    created_at: datetime
    executed_at: datetime | None


class AccountClonePlanOut(ApiModel):
    id: int
    tenant_id: int
    source_account_id: int
    target_account_id: int | None = None
    target_account_ids: list[int] = []
    target_accounts_summary: list[dict[str, Any]] = []
    clone_scope: str
    status: str
    items_total: int
    items_done: int
    items_failed: int
    failure_detail: str
    created_by: str
    created_at: datetime
    confirmed_at: datetime | None
    items: list[AccountCloneItemOut] = []
    items_by_target: dict[str, list[AccountCloneItemOut]] = {}


class RecommendedAccountOut(BaseModel):
    group_id: int
    group_title: str
    account_id: int
    account_name: str
    username: str | None
    health_score: float
    can_send: bool
    is_selectable: bool = True
    unavailable_reason: str | None = None
    cooldown_until: datetime | None = None
    recommended: bool
    reason: str


# ── Composite detail outputs ──

class AccountRiskDiagnosticOut(BaseModel):
    level: str
    code: str
    title: str
    detail: str
    source: str
    action: str
    occurred_at: datetime | None = None


class AccountDetailOut(BaseModel):
    account: AccountOut
    risk_diagnostics: list[AccountRiskDiagnosticOut] = []
    login_flows: list[LoginFlowOut]
    verification_codes: list[VerificationCodeOut]
    profile_sync_records: list[ProfileSyncRecordOut]
    sync_records: list[AccountSyncRecordOut] = []
    next_sync_at: datetime | None = None
    sync_due: bool = False
    sync_status_text: str = ""
    contacts: list[ContactOut]
    groups: list[AccountGroupOut]
    operation_targets: list[OperationTargetOut] = []
    message_records: list[MessageTaskOut]
    manual_operation_records: list[ManualOperationRecordOut] = []
    operation_task_attempts: list[OperationTaskAttemptOut] = []
    clone_plans: list[AccountClonePlanOut] = []
    verification_tasks: list[VerificationTaskOut] = []
    stats: dict[str, Any]


class AccountPoolDetailOut(BaseModel):
    pool: AccountPoolOut
    accounts: list[AccountOut]
    contacts: list[ContactOut]
    verification_tasks: list[VerificationTaskOut] = []
    clone_plans: list[AccountClonePlanOut] = []
    message_records: list[MessageTaskOut] = []
    stats: dict[str, Any]


__all__ = [
    "TgAccountCreate", "AccountPoolCreate", "AccountPoolUpdate",
    "MoveAccountPoolRequest", "TgAccountProfileUpdate",
    "LoginStartRequest", "LoginVerifyRequest", "SensitiveActionReasonRequest", "AccountClonePlanCreate",
    "AvatarUploadOut",
    "AccountOut", "AccountPoolOut", "LoginFlowOut", "VerificationCodeOut",
    "ProfileSyncRecordOut", "AccountSyncRecordOut",
    "ContactOut", "AccountGroupOut",
    "AccountCloneItemOut", "AccountClonePlanOut",
    "RecommendedAccountOut", "AccountRiskDiagnosticOut",
    "AccountDetailOut", "AccountPoolDetailOut",
]
